### PR TITLE
Add GPG signature verification to install script

### LINKED ISSUE #7838: Binary releases content trust
Output of `helm version`: `2.16.5`

Output of `kubectl version`: no need of any `kubectl` infos to accurately describe my issue, plus it would not make sense to.

Cloud Provider/Platform (AKS, GKE, Minikube etc.): same as for `kubectl`

> of

Hi Helm Team, my issue is about installing helm, sha256sums of downloadable released binaries are corrupted or not managed : 

```bash
jbl@poste-devops-jbl-16gbram:~/siecle$ curl -LO https://get.helm.sh/helm-v2.16.5-linux-amd64.tar.gz.sha256
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    65  100    65    0     0     86      0 --:--:-- --:--:-- --:--:--    86
jbl@poste-devops-jbl-16gbram:~/siecle$ cat helm-v2.16.5-linux-amd64.tar.gz.sha256 
7cd54f56f787d62fd7311d5ae2c26e31310d37d27e3622014e5c769bec008cab
jbl@poste-devops-jbl-16gbram:~/siecle$ cat checksumsjbl 
7cd54f56f787d62fd7311d5ae2c26e31310d37d27e3622014e5c769bec008cab ./helm-v2.16.5-linux-amd64.tar.gz.sha256
jbl@poste-devops-jbl-16gbram:~/siecle$ sha256sum 
^C
jbl@poste-devops-jbl-16gbram:~/siecle$ sha256sum -c ./helm-v2.16.5-linux-amd64.tar.gz
sha256sum: ./helm-v2.16.5-linux-amd64.tar.gz: no properly formatted SHA256 checksum lines found
jbl@poste-devops-jbl-16gbram:~/siecle$ sha256sum -c ./checksumsjbl 
./helm-v2.16.5-linux-amd64.tar.gz.sha256: FAILED
sha256sum: WARNING: 1 computed checksum did NOT match
jbl@poste-devops-jbl-16gbram:~/siecle$ ls -allg
total 24708
drwxr-xr-x  2 jbl     4096 Mar 29 17:43 .
drwxr-xr-x 97 jbl    12288 Mar 29 17:36 ..
-rw-r--r--  1 jbl      106 Mar 29 17:41 checksumsjbl
-rw-r--r--  1 jbl 25273394 Mar 29 17:41 helm-v2.16.5-linux-amd64.tar.gz
-rw-r--r--  1 jbl       65 Mar 29 17:43 helm-v2.16.5-linux-amd64.tar.gz.sha256
jbl@poste-devops-jbl-16gbram:~/siecle$ 
```

So now I am bound to build helm from source to get a production grade secure installation of Helm.

Additionnaly, the build from source documentation is not obviously found. That is really unfriendly, if not surprising. 

As far as I can get, it looks like you are at the "just get the source and check my GPG signature on commits and or tags" level. 

I would gladly propose my contribution on providing documentation management into helm project, if desired.

### LINKED ISSUE #7943: Install script should have a way to verify release signatures
Related to conversation in https://github.com/helm/helm/issues/7838

### PR DESCRIPTION
**Pull request by jdolitsky**
The script fetches the KEYS file from GitHub, as well as the .asc files on the release and verifies the
release artifacts are signed by a valid key.

Added new boolean config options in the install script which allow for fine-grained control over verification and output:

- DEBUG: sets -x in the bash script (default: false)
- VERIFY_CHECKSUM: verifies checksum (default: true)
- VERIFY_SIGNATURE: verifies signature (default: true)

Also reduced check for curl/wget to only one time.

Resolves #7943.
Resolves #7838.

### CODE REVIEW

**Comment by jdolitsky** [comment]
I'm wondering if I should push the GPG output into /dev/null? Turns the output of the script from just 3 lines to ~30 and displays all the maintainers emails ... maybe only if DEBUG is enabled, show GPG output?

File: scripts/get-helm-3 [hunk 1]
```diff
@@ -19,7 +19,16 @@
 
 : ${BINARY_NAME:="helm"}
 : ${USE_SUDO:="true"}
+: ${DEBUG:="false"}
+: ${VERIFY_CHECKSUM:="true"}
+: ${VERIFY_SIGNATURES:="true"}
```
[the remark this answers is not in the document]
Comment (jdolitsky) [reply scripts/get-helm-3]: Yea, I know... I've made it disabled by default. I'd rather get this merged with least the ability to verify. We can always submit a 1 line PR to enable verification by default. At the very least we can add a note in the docs.

**Comment by jdolitsky** [comment]
I've silenced the GPG output unless DEBUG=true, now the output looks like this:
```
$ VERIFY_SIGNATURES=true scripts/get-helm-3
Downloading https://get.helm.sh/helm-v3.1.2-darwin-amd64.tar.gz
Verifying checksum... Done.
Verifying signatures... Done.
Preparing to install helm into /usr/local/bin
helm installed into /usr/local/bin/helm
```

File: scripts/get-helm-3 [hunk 2]
```diff
@@ -115,35 +123,97 @@ downloadFile() {
...
+  elif [ "${HAS_WGET}" == "true" ]; then
+    wget -q -O "$HELM_SUM_FILE" "$CHECKSUM_URL"
     wget -q -O "$HELM_TMP_FILE" "$DOWNLOAD_URL"
   fi
 }
 
-# installFile verifies the SHA256 for the file, then unpacks and
-# installs it.
+# verifyFile verifies the SHA256 checksum of the binary package
+# and the GPG signtaures for both the package and checksum file
```
Comment (bacongobbler) [inline scripts/get-helm-3:136]: typo: signatures

File: scripts/get-helm-3 [hunk 3]
```diff
@@ -115,35 +123,97 @@ downloadFile() {
...
+  echo "Done."
+}
 
-  mkdir -p "$HELM_TMP"
-  tar xf "$HELM_TMP_FILE" -C "$HELM_TMP"
-  HELM_TMP_BIN="$HELM_TMP/$OS-$ARCH/helm"
-  echo "Preparing to install $BINARY_NAME into ${HELM_INSTALL_DIR}"
-  runAsRoot cp "$HELM_TMP_BIN" "$HELM_INSTALL_DIR/$BINARY_NAME"
-  echo "$BINARY_NAME installed into $HELM_INSTALL_DIR/$BINARY_NAME"
+# verifySignatures obtains the KEYS and signature .asc files from GitHub,
```
Comment (bacongobbler) [inline scripts/get-helm-3:175]: Is it safe to fetch the public keys from the internet in an automated fashion? Shouldn't this be performed out-of-band, like how you have to do with `apt-key add` before a `apt update`? That way the user can verify and trust the imported keys before running the script.

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: scripts/get-helm-3 (hunk 3 above)
Comment (jdolitsky) [reply scripts/get-helm-3:175]: I was chatting w/ @mattfarina about this - my original thought was you could give it the path to a .asc/.pub file that had already been downloaded (from keybase etc) and we use that to import. The third option is take out the `gpg --import`  altogether (which I think is what you are suggesting?). I'm impartial to the approach here

File: scripts/get-helm-3 (hunk 2 above)
Comment (jdolitsky) [reply scripts/get-helm-3]: thanks! fixed

File: scripts/get-helm-3 (hunk 3 above)
Comment (jdolitsky) [reply scripts/get-helm-3]: I've gone ahead and removed the import altogether

**Comment by jdolitsky** [comment]
Modified the script. No longer auto-imports KEYS file, just expects you've done this prior to running the script.

File: scripts/get-helm-3 [hunk 4]
```diff
@@ -115,35 +123,97 @@ downloadFile() {
...
+  if [ "${HAS_CURL}" == "true" ]; then
+    curl -SsL "${github_keys_url}" -o "${HELM_TMP_ROOT}/${keys_filename}"
+  elif [ "${HAS_WGET}" == "true" ]; then
+    wget -q -O "${github_keys_url}" "${HELM_TMP_ROOT}/${keys_filename}"
+  fi
+  local gpg_stderr_device="/dev/null"
+  if [ "${DEBUG}" == "true" ]; then
+    gpg_stderr_device="/dev/stderr"
+  fi
+  gpg --import "${HELM_TMP_ROOT}/${keys_filename}" 2> "${gpg_stderr_device}"
```
Comment (mattfarina) [inline scripts/get-helm-3]: This is going to add the keys to the users active GPG keyring. The gpg verification below will use these keys plus all the other keys in the keyring to look for a good signature.

The gpg keys from the Helm project should not be imported into the users keyring. Instead, a keyring only containing the Helm maintainers should be created, used for verification, and deleted afterwards as part of cleanup. This script should not try to use keys in the users existing keyring and should not alter their environment.

The linux kernel script does this. You can look at it for ideas... https://git.kernel.org/pub/scm/linux/kernel/git/mricon/korg-helpers.git/tree/get-verified-tarball

**Review by mattfarina — CHANGES_REQUESTED** [review]

File: scripts/get-helm-3 [hunk 5]
```diff
@@ -19,8 +19,16 @@
 
 : ${BINARY_NAME:="helm"}
 : ${USE_SUDO:="true"}
+: ${DEBUG:="false"}
+: ${VERIFY_CHECKSUM:="true"}
+: ${VERIFY_SIGNATURES:="false"}
 : ${HELM_INSTALL_DIR:="/usr/local/bin"}
 
+HAS_CURL="$(type "curl" > /dev/null && echo true || echo false)"
+HAS_WGET="$(type "wget" > /dev/null && echo true || echo false)"
```
Comment (bacongobbler) [inline scripts/get-helm-3:28]: I found this in my terminal while testing this script:

```
bash: line 28: type: wget: not found
```

If you change this to `&> /dev/null` for all `HAS_*` variables, this output will be silenced.

File: scripts/get-helm-3 (hunk 4 above)
Comment (jdolitsky) [reply scripts/get-helm-3]: > This script should not try to use keys in the users existing keyring and should not alter their environment.

Alright. Thanks for the example script.

So then, should we modify this to take a filepath to a .asc file? Or auto-fetch KEYS file?

**Review by bacongobbler — CHANGES_REQUESTED** [review]
One more nitpick, then this looks good!

```
><> export VERIFY_SIGNATURES=true
><> ./get-helm-3 
Helm v3.2.0 is available. Changing from version v3.1.2.
Downloading https://get.helm.sh/helm-v3.2.0-darwin-amd64.tar.gz
Verifying checksum... Done.
Verifying signatures... Done.
Preparing to install helm into /usr/local/bin
Password: ^C
Failed to install helm
	For support, go to https://github.com/helm/helm.
```

File: scripts/get-helm-3 (hunk 4 above)
Comment (jdolitsky) [reply scripts/get-helm-3]: I think what you're saying is at odds with @bacongobbler 's suggestion?

File: scripts/get-helm-3 (hunk 5 above)
Comment (jdolitsky) [reply scripts/get-helm-3]: You dont have wget ?? :)
Thanks, resolved

File: scripts/get-helm-3 (hunk 4 above)
Comment (bacongobbler) [reply scripts/get-helm-3]: yes and no. I wasn't aware that the process was to create a separate keyring as opposed to using your existing keyring.

If that's the recommendation level at the linux kernel, then that should work for us.

Comment (mattfarina) [reply scripts/get-helm-3]: It's possible to have a separate keyring used for gpg verification calls. It's a couple steps to set it up. We should make it so the only keys in the keyring are those from the maintainers. This way the good verification can only come from that set of keys and not any key in the users personal keyring.

**Comment by jdolitsky** [comment]
@bacongobbler @mattfarina this should be good to go now.

Running into an issue on Mac, which I will file once this is closed:
```
gpg: can't connect to the agent: IPC connect call failed
```

Appears to be fixed if the temp gpg home is somewhere under the users home? The ownership on the temp dir seem find.. Found some conversation about it here: https://sourceforge.net/p/macgpg/mailman/message/18850673/

So for now, I've put a check in place which requires Linux if using VERIFY_SIGNATURES=true (false by default)

For proof this isnt messing with the user's active keyring:
```
$ gpg --list-keys | grep ^pub | wc -l
0

$ curl -s https://keybase.io/jdolitsky/pgp_keys.asc | gpg --import 2>/dev/null

$ gpg --list-keys | grep ^pub | wc -l
1

$ VERIFY_SIGNATURES=true scripts/get-helm-3
Downloading https://get.helm.sh/helm-v3.2.0-linux-amd64.tar.gz
Verifying checksum... Done.
Verifying signatures... Done.
Preparing to install helm into /usr/local/bin
helm installed into /usr/local/bin/helm

$ gpg --list-keys | grep ^pub | wc -l
1
```

File: scripts/get-helm-3 [hunk 6]
```diff
@@ -115,35 +142,94 @@ downloadFile() {
...
+    echo -e "${error_text}"
+    exit 1
+  fi
+  local num_goodlines_tar=$(gpg --verify --keyring="${gpg_keyring}" --status-fd=1 "${HELM_TMP_ROOT}/helm-${TAG}-${OS}-${ARCH}.tar.gz.asc" 2> "${gpg_stderr_device}" | grep -c -E '^\[GNUPG:\] (GOODSIG|VALIDSIG)')
+  if [[ ${num_goodlines_tar} -lt 2 ]]; then
+    echo "Unable to verify the signature of helm-${TAG}-${OS}-${ARCH}.tar.gz!"
+    echo -e "${error_text}"
+    exit 1
+  fi
+  echo "Done."
```
Comment (jdolitsky) [inline scripts/get-helm-3:232]: this strange grep + line count check was taken directly from the linux kernel script https://git.kernel.org/pub/scm/linux/kernel/git/mricon/korg-helpers.git/tree/get-verified-tarball#n164

**Comment by jdolitsky** [comment]
@mattfarina @bacongobbler want to try to get this into the install script for 3.3?

File: scripts/get-helm-3 [hunk 7]
```diff
@@ -67,20 +75,39 @@ verifySupported() {
...
+  fi
+
+  if [ "${VERIFY_SIGNATURES}" == "true" ]; then
+    if [ "${HAS_GPG}" != "true" ]; then
+      echo "In order to verify signatures, gpg must first be installed."
+      echo "Please install gpg or set VERIFY_SIGNATURES=false in your environment."
+      exit 1
+    fi
+    if [ "${OS}" != "linux" ]; then
+      echo "Signature verification is currently only supported on Linux."
```
Comment (bacongobbler) [inline scripts/get-helm-3:96]: why only linux support when there's native gpg support for MacOS? Can you please elaborate what needs to change to add MacOS support?

**Review by bacongobbler — CHANGES_REQUESTED** [review]
does not work. Please test your changes before asking for a review.

```
><> docker run -it ubuntu bash
root@c5a6147940e8:/# apt update && apt install wget git openssl gpg
root@c5a6147940e8:/# wget https://raw.githubusercontent.com/jdolitsky/helm/validate-signature/scripts/get-helm-3
root@c5a6147940e8:/# chmod 755 get-helm-3
root@c5a6147940e8:/# VERIFY_SIGNATURES=true ./get-helm-3 
Downloading https://get.helm.sh/helm-v3.2.4-linux-amd64.tar.gz
Verifying checksum... Done.
Verifying signatures... https://raw.githubusercontent.com/helm/helm/master/KEYS: No such file or directory
Failed to install helm
	For support, go to https://github.com/helm/helm.
```

**Comment by jdolitsky** [comment]
@bacongobbler - apologies. Was testing on a Linux system with curl for the 2nd revision and didn't catch the bad syntax with wget commands. I just updated and tested it on system with just curl, and system with just wget

Regarding Mac support - it's not clear to me what the issue is. Here is the output when using `DEBUG=true`:
```
...
+ gpg --batch --quiet --homedir=/var/folders/lj/xzzflpn14bg1vd8f2c5q_v500000gn/T/helm-installer-XXXXXX.JjFTM6Vw/gnupg --import /var/folders/lj/xzzflpn14bg1vd8f2c5q_v500000gn/T/helm-installer-XXXXXX.JjFTM6Vw/KEYS
gpg: can't connect to the agent: IPC connect call failed
gpg: can't connect to the agent: IPC connect call failed
gpg: can't connect to the agent: IPC connect call failed
gpg: can't connect to the agent: IPC connect call failed
gpg: can't connect to the agent: IPC connect call failed
gpg: can't connect to the agent: IPC connect call failed
+ fail_trap
+ result=2
+ '[' 2 '!=' 0 ']'
+ [[ -n '' ]]
+ echo 'Failed to install helm'
Failed to install helm
...
```

It appears to hang on `can't connect to the agent`. Was unable to successfully find something on the internet to resolve this without launching a GPG agent (which I don't think the script should do).

If I modify `HELM_TMP_ROOT` to be instead `${HOME}/helm-install` (vs. `$(mktemp -dt helm-installer-XXXXXX)`) on line 141, things seem to work just fine:

```
...
+ gpg --batch --quiet --homedir=/Users/jdolitsky/helm-install/gnupg --import /Users/jdolitsky/helm-install/KEYS
+ gpg --batch --no-default-keyring --keyring /Users/jdolitsky/helm-install/gnupg/pubring.kbx --export
+ local github_release_url=https://github.com/helm/helm/releases/download/v3.2.4
+ '[' true == true ']'
+ curl -SsL https://github.com/helm/helm/releases/download/v3.2.4/helm-v3.2.4-darwin-amd64.tar.gz.sha256.asc -o /Users/jdolitsky/helm-install/helm-v3.2.4-darwin-amd64.tar.gz.sha256.asc
...
```

So it's something related to using `mktemp`. When I inspect these temp directories, they appear to be fully owned by my user:

```
+ HELM_TMP_ROOT=/var/folders/lj/xzzflpn14bg1vd8f2c5q_v500000gn/T/helm-installer-XXXXXX.hGNuwU8u
+ ls -la /var/folders/lj/xzzflpn14bg1vd8f2c5q_v500000gn/T/helm-installer-XXXXXX.hGNuwU8u
total 0
drwx------    2 jdolitsky  staff    64 Jul  7 13:08 .
drwx------@ 136 jdolitsky  staff  4352 Jul  7 13:08 ..
```

I could definitely use some help figuring this out. But I do think this PR still provides value for Linux users, especially in the context of CI etc.

**Review by bacongobbler — APPROVED** [review]
re-tested using the same instructions. Works now. I think this is in a workable state to move forward. We can address the MacOS issue in a follow-up.

**Pull request merged**